In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

In [2]:
train_df = pd.read_csv(Path('2019loans.csv'))
test_df = pd.read_csv(Path('2020Q1loans.csv'))

In [3]:
train_df.head()

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   

   annual_inc verification_status loan_status pymnt_plan  ...  pct_tl_nvr_dlq  \
0    223000.0        Not Verified    low_risk          n  ...           100.0   
1    123000.0     Source Verified    low_risk          n  ...            85.0   
2    197000.0     Source Verified    low_risk          n  ...            85.7   
3     45000.0        Not Verified    low_risk          n  ...           100.0   
4    133000.0     Source Verified    low_risk          n  ...           100.0   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              50.0                   0.0        0.0         577150.0   
1              33.3                   0.0        0.0         132750.0   
2              33.3                   0.0        0.0         628160.0   
3              16.7                   1.0        0.0          42006.0   
4              66.7                   0.0        0.0         283248.0   

   total_bal_ex_mort  total_bc_limit total_il_high_credit_limit  \
0           122018.0         32000.0                   170200.0   
1            27896.0         15900.0                    35398.0   
2           114043.0         22600.0                    90340.0   
3            20761.0         19900.0                    15406.0   
4           109056.0         79500.0                    58778.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  
2              N                     N  
3              N                     N  
4              N                     N  

[5 rows x 86 columns]

In [4]:
# pd.get_dummies()
train_df["y"]= pd.get_dummies(train_df["loan_status"],drop_first=True ).rename(columns={"low_risk":"y"})


In [5]:
train_df.drop(columns=["loan_status", "Unnamed: 0", "index"], inplace=True)

In [6]:
# pd.get_dummies()
test_df["y"]= pd.get_dummies(test_df["loan_status"],drop_first=True ).rename(columns={"low_risk":"y"})


In [7]:
test_df.drop(columns=["loan_status", "Unnamed: 0", "index"], inplace=True)

In [ ]:
train_df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    13375.0    0.1797       483.34       MORTGAGE    223000.0   
1    21000.0    0.1308       478.68       MORTGAGE    123000.0   
2    20000.0    0.1240       448.95       MORTGAGE    197000.0   
3     3000.0    0.1240       100.22           RENT     45000.0   
4    30000.0    0.1612      1056.49       MORTGAGE    133000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  29.99          0.0             0.0  ...   
1     Source Verified          n  11.26          2.0             0.0  ...   
2     Source Verified          n  11.28          0.0             0.0  ...   
3        Not Verified          n  18.08          0.0             0.0  ...   
4     Source Verified          n  27.77          0.0             2.0  ...   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              50.0                   0.0        0.0         577150.0   
1              33.3                   0.0        0.0         132750.0   
2              33.3                   0.0        0.0         628160.0   
3              16.7                   1.0        0.0          42006.0   
4              66.7                   0.0        0.0         283248.0   

  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0          122018.0         32000.0                    170200.0   
1           27896.0         15900.0                     35398.0   
2          114043.0         22600.0                     90340.0   
3           20761.0         19900.0                     15406.0   
4          109056.0         79500.0                     58778.0   

   hardship_flag  debt_settlement_flag  y  
0              N                     N  1  
1              N                     N  1  
2              N                     N  1  
3              N                     N  1  
4              N                     N  1  

[5 rows x 84 columns]

In [8]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12180 entries, 0 to 12179
Data columns (total 84 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   loan_amnt                   12180 non-null  float64
 1   int_rate                    12180 non-null  float64
 2   installment                 12180 non-null  float64
 3   home_ownership              12180 non-null  object 
 4   annual_inc                  12180 non-null  float64
 5   verification_status         12180 non-null  object 
 6   pymnt_plan                  12180 non-null  object 
 7   dti                         12180 non-null  float64
 8   delinq_2yrs                 12180 non-null  float64
 9   inq_last_6mths              12180 non-null  float64
 10  open_acc                    12180 non-null  float64
 11  pub_rec                     12180 non-null  float64
 12  revol_bal                   12180 non-null  float64
 13  total_acc                   121

In [9]:
train_df.select_dtypes(include="object")

home_ownership verification_status pymnt_plan initial_list_status  \
0           MORTGAGE        Not Verified          n                   w   
1           MORTGAGE     Source Verified          n                   w   
2           MORTGAGE     Source Verified          n                   w   
3               RENT        Not Verified          n                   w   
4           MORTGAGE     Source Verified          n                   w   
...              ...                 ...        ...                 ...   
12175           RENT        Not Verified          n                   w   
12176           RENT            Verified          n                   w   
12177           RENT        Not Verified          n                   w   
12178       MORTGAGE     Source Verified          n                   w   
12179           RENT        Not Verified          n                   f   

      application_type hardship_flag debt_settlement_flag  
0           Individual             N                    N  
1           Individual             N                    N  
2           Individual             N                    N  
3           Individual             N                    N  
4           Individual             N                    N  
...                ...           ...                  ...  
12175       Individual             N                    N  
12176       Individual             N                    N  
12177       Individual             N                    N  
12178        Joint App             N                    N  
12179       Individual             N                    N  

[12180 rows x 7 columns]

In [10]:
pd.get_dummies(train_df.select_dtypes(include="object"))

home_ownership_ANY  home_ownership_MORTGAGE  home_ownership_OWN  \
0                       0                        1                   0   
1                       0                        1                   0   
2                       0                        1                   0   
3                       0                        0                   0   
4                       0                        1                   0   
...                   ...                      ...                 ...   
12175                   0                        0                   0   
12176                   0                        0                   0   
12177                   0                        0                   0   
12178                   0                        1                   0   
12179                   0                        0                   0   

       home_ownership_RENT  verification_status_Not Verified  \
0                        0                                 1   
1                        0                                 0   
2                        0                                 0   
3                        1                                 1   
4                        0                                 0   
...                    ...                               ...   
12175                    1                                 1   
12176                    1                                 0   
12177                    1                                 1   
12178                    0                                 0   
12179                    1                                 1   

       verification_status_Source Verified  verification_status_Verified  \
0                                        0                             0   
1                                        1                             0   
2                                        1                             0   
3                                        0                             0   
4                                        1                             0   
...                                    ...                           ...   
12175                                    0                             0   
12176                                    0                             1   
12177                                    0                             0   
12178                                    1                             0   
12179                                    0                             0   

       pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0                 1                      0                      1   
1                 1                      0                      1   
2                 1                      0                      1   
3                 1                      0                      1   
4                 1                      0                      1   
...             ...                    ...                    ...   
12175             1                      0                      1   
12176             1                      0                      1   
12177             1                      0                      1   
12178             1                      0                      1   
12179             1                      1                      0   

       application_type_Individual  application_type_Joint App  \
0                                1                           0   
1                                1                           0   
2                                1                           0   
3                                1                           0   
4                                1                           0   
...                            ...                         ...   
12175                            1                           0   
12176                            1                           0   
12177        

In [11]:
train_df.select_dtypes(exclude="object")

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0        13375.0    0.1797       483.34    223000.0  29.99          0.0   
1        21000.0    0.1308       478.68    123000.0  11.26          2.0   
2        20000.0    0.1240       448.95    197000.0  11.28          0.0   
3         3000.0    0.1240       100.22     45000.0  18.08          0.0   
4        30000.0    0.1612      1056.49    133000.0  27.77          0.0   
...          ...       ...          ...         ...    ...          ...   
12175    19975.0    0.2565       801.09     28000.0  28.42          0.0   
12176    15000.0    0.1774       540.34     50000.0  23.43          4.0   
12177     3600.0    0.1862       131.28     60000.0  28.80          0.0   
12178    15000.0    0.0881       475.68     62000.0  11.44          0.0   
12179    15000.0    0.1774       540.34     54080.0  21.66          1.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  num_tl_op_past_12m  \
0                 0.0      15.0      0.0    39728.0  ...                 5.0   
1                 0.0      16.0      0.0     9585.0  ...                 0.0   
2                 0.0      12.0      0.0    16708.0  ...                 1.0   
3                 0.0      12.0      1.0     8809.0  ...                 4.0   
4                 2.0      13.0      0.0    65420.0  ...                 2.0   
...               ...       ...      ...        ...  ...                 ...   
12175             0.0      15.0      0.0     9679.0  ...                 3.0   
12176             0.0      16.0      0.0     3193.0  ...                 2.0   
12177             1.0      14.0      0.0    11804.0  ...                 5.0   
12178             0.0       5.0      0.0      157.0  ...                 1.0   
12179             1.0       5.0      0.0     8349.0  ...                 1.0   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               100.0              50.0                   0.0        0.0   
1                85.0              33.3                   0.0        0.0   
2                85.7              33.3                   0.0        0.0   
3               100.0              16.7                   1.0        0.0   
4               100.0              66.7                   0.0        0.0   
...               ...               ...                   ...        ...   
12175           100.0              16.7                   0.0        0.0   
12176            90.5              11.1                   0.0        0.0   
12177           100.0               0.0                   0.0        0.0   
12178           100.0               0.0                   0.0        0.0   
12179            90.0             100.0                   0.0        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0             577150.0           122018.0         32000.0   
1             132750.0            27896.0         15900.0   
2             628160.0           114043.0         22600.0   
3              42006.0            20761.0         19900.0   
4             283248.0           109056.0         79500.0   
...                ...                ...             ...   
12175          50055.0            28192.0         18700.0   
12176          70324.0            57025.0         13300.0   
12177          83765.0            55156.0         14800.0   
12178         189930.0            23748.0          7000.0   
12179          48968.0            29786.0          2900.0   

       total_il_high_credit_limit  y  
0                        170200.0  1  
1                         35398.0  1  
2                         90340.0  1  
3                         15406.0  1  
4                         58778.0  1  
...                           ... ..  
12175                     19055.0  0  
12176                     54824.0  0  
12177                     53065.0  0  
12178                     32930.0  0  
12179                     33968.0  0  

[12180 rows x 77 columns]

In [12]:
pd.concat([train_df.select_dtypes(exclude="object"),pd.get_dummies(train_df.select_dtypes(include="object"))], axis=1)

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0        13375.0    0.1797       483.34    223000.0  29.99          0.0   
1        21000.0    0.1308       478.68    123000.0  11.26          2.0   
2        20000.0    0.1240       448.95    197000.0  11.28          0.0   
3         3000.0    0.1240       100.22     45000.0  18.08          0.0   
4        30000.0    0.1612      1056.49    133000.0  27.77          0.0   
...          ...       ...          ...         ...    ...          ...   
12175    19975.0    0.2565       801.09     28000.0  28.42          0.0   
12176    15000.0    0.1774       540.34     50000.0  23.43          4.0   
12177     3600.0    0.1862       131.28     60000.0  28.80          0.0   
12178    15000.0    0.0881       475.68     62000.0  11.44          0.0   
12179    15000.0    0.1774       540.34     54080.0  21.66          1.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                 0.0      15.0      0.0    39728.0  ...   
1                 0.0      16.0      0.0     9585.0  ...   
2                 0.0      12.0      0.0    16708.0  ...   
3                 0.0      12.0      1.0     8809.0  ...   
4                 2.0      13.0      0.0    65420.0  ...   
...               ...       ...      ...        ...  ...   
12175             0.0      15.0      0.0     9679.0  ...   
12176             0.0      16.0      0.0     3193.0  ...   
12177             1.0      14.0      0.0    11804.0  ...   
12178             0.0       5.0      0.0      157.0  ...   
12179             1.0       5.0      0.0     8349.0  ...   

       verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                 0             1                      0   
1                                 0             1                      0   
2                                 0             1                      0   
3                                 0             1                      0   
4                                 0             1                      0   
...                             ...           ...                    ...   
12175                             0             1                      0   
12176                             1             1                      0   
12177                             0             1                      0   
12178                             0             1                      0   
12179                             0             1                      1   

       initial_list_status_w  application_type_Individual  \
0                          1                            1   
1                          1                            1   
2                          1                            1   
3                          1                            1   
4                          1                            1   
...                      ...                          ...   
12175                      1                            1   
12176                      1                            1   
12177                      1                            1   
12178                      1                            0   
12179                      0                            1   

       application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                               0                1                0   
1                               0                1                0   
2                               0                1                0   
3                               0                1                0   
4                               0                1                0   
...                           ...              ...              ...   
12175                           0                1                0   
12176                           0                1                0   
12177                           0                1                0   
12178                        

In [13]:
train_df= pd.concat([train_df.select_dtypes(exclude="object"),pd.get_dummies(train_df.select_dtypes(include="object"))], axis=1)

In [14]:
test_df= pd.concat([test_df.select_dtypes(exclude="object"),pd.get_dummies(test_df.select_dtypes(include="object"))], axis=1)

In [15]:
train_df.shape

(12180, 93)

In [16]:
test_df.shape

(4702, 92)

In [17]:
for x in train_df.columns: 
  if x not in test_df.columns:
    print(x)
    test_df[x]=0

debt_settlement_flag_Y


In [18]:
test_df.shape

(4702, 93)

In [20]:
y_train=train_df["y"]
X_train=train_df.drop(columns=["y"])

In [21]:
y_test=test_df["y"]
X_test=test_df.drop(columns=["y"])

In [31]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(max_iter=4000).fit(X_train, y_train)
clf.score(X_test, y_test)

print(f"Training Data Score Logistic Regression: {clf.score(X_train, y_train)}")
print(f"Testing Data Score Logistic Regression: {clf.score(X_test, y_test)}")


Training Data Score Logistic Regression: 0.7029556650246306
Testing Data Score Logistic Regression: 0.5776265418970651


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


## Logistic Regression model on the unscaled data
I got a better Testing Score using max_inter=4000 (maximum number of iterations taken for the solvers to converge) , having a Testing Data Score: 0.5776265418970651. I got 0.51.... without max_iner 

In [30]:
# Train a Random Forest Classifier model and print the model score

from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state=0)
clf.fit(X_train, y_train)
clf.score(X_test, y_test)

print(f"Training Data Score Random Forest Clf: {clf.score(X_train, y_train)}")
print(f"Testing Data Score Random Forest Clf: {clf.score(X_test, y_test)}")


Training Data Score Random Forest Clf: 1.0
Testing Data Score Random Forest Clf: 0.6424925563589962


# We got a better result on our TESTING SCORE using the Train Random Forest Classifier, around 10% better result, since fits a number of decision tree classifiers on various sub-samples.

In [32]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_train_scaled

array([[-0.39311205,  0.73658452, -0.08760946, ..., -0.17149859,
         0.02026518, -0.02026518],
       [ 0.35168119, -0.19171582, -0.10342722, ..., -0.17149859,
         0.02026518, -0.02026518],
       [ 0.25400339, -0.32080462, -0.20434179, ..., -0.17149859,
         0.02026518, -0.02026518],
       ...,
       [-1.34791257,  0.85997823, -1.28263075, ..., -0.17149859,
         0.02026518, -0.02026518],
       [-0.23438563, -1.00231755, -0.11361032, ..., -0.17149859,
         0.02026518, -0.02026518],
       [-0.23438563,  0.69292214,  0.10586953, ..., -0.17149859,
         0.02026518, -0.02026518]])

In [33]:
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[ 2.20755943, -1.12001617,  1.0371484 , ..., -0.17149859,
         0.02026518, -0.02026518],
       [-1.11348584,  0.21833096, -1.01983876, ..., -0.17149859,
         0.02026518, -0.02026518],
       [-1.34791257,  0.54295132, -1.2928478 , ..., -0.17149859,
         0.02026518, -0.02026518],
       ...,
       [-0.72277464,  1.7009538 , -0.41340093, ..., -0.17149859,
         0.02026518, -0.02026518],
       [-0.91813024,  0.85997823, -1.02947877, ..., -0.17149859,
         0.02026518, -0.02026518],
       [ 1.23078141,  1.22636262,  2.08478621, ..., -0.17149859,
         0.02026518, -0.02026518]])

In [37]:
# Train the Logistic Regression model on the scaled data and print the model score

clf = LogisticRegression(max_iter=4000).fit(X_train_scaled, y_train)
clf.score(X_test_scaled, y_test)

print(f"Training Data Score Log. Regression SCALE DATA: {clf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score Log. Regression SCALE DATA: {clf.score(X_test_scaled, y_test)}")


Training Data Score Log. Regression SCALE DATA: 0.7079638752052545
Testing Data Score Log. Regression SCALE DATA: 0.7677584006805614


## Using the SCALED DATA the Logistic Regression have aroun 20% better result than getting the Logostic Regression without scale the data. 

In [38]:
# Train a Random Forest Classifier model on the scaled data and print the model score

from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state=0)
clf.fit(X_train_scaled, y_train)
clf.score(X_test_scaled, y_test)

print(f"Training Data Score Random Forest SCALE DATA: {clf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score Random Forest SCALE DATA: {clf.score(X_test_scaled, y_test)}")


Training Data Score Random Forest SCALE DATA: 1.0
Testing Data Score Random Forest SCALE DATA: 0.6437686091025095


# Random Forest using the SCALED DATA doesn't have a better result than not using the scale data. 
